---
title: "Agilent Single-Channel Microarray Processing Pipeline RMarkdown (GLDS-156)"
output: html_document
date: "`r Sys.Date()`"
---

## Installing necessary packages

For processing Agilent microarray data, I needed to install a few packages that were used for the importing and analyzing the data. 
This included Bioconductor as well as Linear Models for Microarray (Limma). These are popular packages for processing microarray data.

```{r}
## For installing Bioconductor on R
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install(version = "3.14")

## For installing limma on R
# if (!require("BiocManager", quietly = TRUE))
#     install.packages("BiocManager")
# BiocManager::install("limma")

##if (!require("BiocManager", quietly = TRUE))
#   install.packages("BiocManager")
#
#BiocManager::install("biomaRt")
```
## Load the packages
After installing the packages above, I had to load them in using the library function so that I could access their functions.
```{r}
library(limma)
library(dplyr)
library(statmod)
library(biomaRt)
```
## Importing the Raw Data
Now that the library was loaded, I was able to import the raw data. I started by importing the runsheet into R.
This contained the necessary data for processing later in the pipeline.
```{r Importing_runsheets}
dir = "C:/Users/tangk/Desktop/Agilent_Microarray"
GLDS_156_rs <- read.csv(file.path(dir, "Runsheet(GLDS-156).csv"), check.names = FALSE, fileEncoding = 'UTF-8-BOM') ## Outputs a dataframe
Factor_Value <- GLDS_156_rs[,c("Factor Value[Dose]", "Factor Value[Ionizing Radiation]")]

compare_csv_from_runsheet <- function(runsheet_path) {
    df = read.csv(runsheet_path, check.names = FALSE, fileEncoding = 'UTF-8-BOM') # fileEncoding removes strange characters from the column names
    # get only Factor Value columns
    factors = as.data.frame(df[,grep("Factor Value", colnames(df), ignore.case=TRUE)])
    colnames(factors) = paste("factor",1:dim(factors)[2], sep= "_")
    result = data.frame(sample_id = df[,c("Sample Name")], factors)	
    return(result)
}
compare_csv <- compare_csv_from_runsheet(file.path(dir, "Runsheet(GLDS-156).csv"))
study <- as.data.frame(compare_csv[,2:dim(compare_csv)[2]])
colnames(study) <- colnames(compare_csv)[2:dim(compare_csv)[2]]
rownames(study) <- compare_csv[,1]
#DT::datatable(study, caption = "TBA")

##### Format groups and indicate the group that each sample belongs to #####
if (dim(study) >= 2){
    group<-apply(study,1,paste,collapse = " & ") # concatenate multiple factors into one condition per sample
} else{
    group<-study[,1]
}
group_names <- paste0("(",group,")",sep = "") # human readable group names
group <- make.names(group) # group naming compatible with R models
names(group) <- group_names
#DT::datatable(as.data.frame(group), caption = "TBA")
levels <- factor(group)
```
- dir = A variable that stores the location of where I kept the runsheets.
- GLDS_156_rs = A varaible that holds the contents of those runsheets.
- Factor_Value = A variable that holds specifically the factor values column from the runsheet.

Here, I imported the raw data from the text files into R using the readmaimages() function from the Limma package.
```{r Raw_Data_Input} 
datadir = "C:/Users/tangk/Desktop/Agilent_Microarray/input_data"
files = dir(path = file.path(datadir, "GLDS-156/GLDS-156_Raw_Data"), pattern="*\\.txt$")

## Input: ".txt" files, Output: EListRaw (Raw expression levels)
raw_data <- limma::read.maimages(files, source="agilent", path = file.path(datadir, "GLDS-156/GLDS-156_Raw_Data"), sep="\t", green.only = TRUE, other.columns = "gIsWellAboveBG")
  
```
-   datadir = A variable that stores the path to the directory that holds the three datasets that I have chosen.
-   files = A variable that will detect all of the text files within the GLDS-156_Raw_Data directory. This is where all of the raw Agilent microarray data is stored for the GLDS-156 dataset.
-   source = "agilent" because the data that I want to import will be agilent microarray data. There are other options for this flag for other types of microarrays
-   path = The path to the directory that holds the dataset's raw data.
-   sep = How the data is separated in those raw data files (in our case it's tab separated).
-   green.only = TRUE, This command is for telling the function that it should only read the green (Cys3) channel rather than red and green since we're only looking at single channel microarrays.

## QC of the Raw Data

In this step, I used some diagnostic functions from the limma package to get a sense of the raw data and to see if there were any abnormalities present.

```{r EListRaw_density_plot, fig.height = 10, fig.width = 10, fig.align = "center"}
# This code block uses the plotDensities() function from limma to generate a graph with the raw intensities vs. density 
plotDensities(raw_data, log = TRUE, legend = "topright" , main = "Density of raw expression values for multiple arrays")
```

- raw_data = EListRaw object containing the raw expression values and information about them
- log = Specifying the densities should be plotted on the log2 scale
- legend = Specifying where the legend should be placed on the graph
- main = title of the graph

```{r Pseudoimage_Plot}
# This code block uses the imageplot() function to generate a pseudoimage for each sample to allow the user to look for any spatial abnormalities in the expressions on the chip.
num_rows <- nrow(raw_data)

find_factors <- function(num) {
    factors_list <- list()
    for (n in 1:num) {
        if((num %% n) == 0) {
            factors_list <- c(factors_list, n)
        }
    }
    return(factors_list)
}
factors <- find_factors(num_rows)

for (sample_name in colnames(raw_data$E)) {
    if (length(raw_data$printer) != 0) {
        imageplot(raw_data$E[,sample_name], layout=raw_data$printer, zlim = c((1.25*min(raw_data$E[,sample_name])),(0.75*max(raw_data$E[,sample_name]))), legend = TRUE, main = sample_name)
        }
    else {
        if (length(factors) %% 2 == 0) {
            rows <- factors[[length(factors)/2]]
            columns <- factors[[length(factors)/2+1]]
            }
        else {
            rows <- factors[[ceiling(length(factors)/2)]]
            columns <- factors[[ceiling(length(factors)/2)]]
            }
        imageplot(raw_data$E[,sample_name], layout=list(ngrid.r = 1, ngrid.c = 1, nspot.r = rows , nspot.c = columns), zlim = c((1.25*min(raw_data$E[,sample_name])),(0.75*max(raw_data$E[,sample_name]))), legend = TRUE, main = sample_name)
    }
} 
```

- raw_data$E[, sample_name] = specifying that we want to use the raw expression values (rows) from each sample (column) as the data for a plot
- layout = using the values from the printer variable from raw_data to specify the dimensions of the pseudoimage generated
- zlim = altering the range of intensities that will be plotted (min = 1.25 min raw expression value, max = 0.75 max raw expression value)
- legend = displays the range of the z values (raw expresison values) 

```{r MA_plot, fig.height = 5, fig.width = 5, fig.align = "center"}
# This code block uses the plotMA() function from limma to generate a MA plot for each of the samples. 
count = 0
for (sample_name in colnames(raw_data$E)) {
    count = count + 1
    suppressWarnings(limma::plotMA(raw_data,array=count,xlab="Average log-expression",ylab="Expression log-ratio(this sample vs. others)", main = sample_name))
}
```

- raw_data = EListRaw object containing the raw intensities and information about them
- array = choosing which array (sample) to use the data from
- xlab = label for the x axis
- ylab = label for the y axis
- main = title for the plot

```{r Foreground_and_Background_Plot}
# This code block uses the plotFB() function from limma to generate a FB plot to compare the foreground and background intensities in each sample. 
count = 0
for (sample_name in colnames(raw_data$E)) {
    count = count + 1
    plotFB(raw_data, array = count, xlab = "log2 Background", ylab = "log2 Foreground", main = sample_name) 
}
```

- raw_data = EListRaw object containing the raw expression values and information about them
- array = specifying which array (sample) to use as the data
- xlab = label for the x axis
- ylab = label for the y axis
- main = title for the plot

## Normalization of the Data
After making sure that the raw data looks good,
I had to remove the technical variation in the raw intensities
that comes from external sources (i.e. Sample preparation)
by performing normalization on the raw data. For this step, I was able to use a couple other
functions from Limma called backgroundCorrect() to subtract the foreground intensities
from the background intensities and normalizeBetweenArrays() to normalize the data. 

```{r Background_correction_and_Normalization}
corrected_data <- limma::backgroundCorrect(raw_data, method = "normexp")
# Input: Background corrected EListRaw, Output: Elist
norm_data <- normalizeBetweenArrays(corrected_data, method = "quantile")  
```

- raw_data = EListRaw object containing the raw intensities and information about the arrays
- method = specifying which background correction method should be used on the raw data
- corrected_data = Background corrected EListRaw object
- method = specifying which normalization method should be done on the data

## QA of the Normalized Data
```{r Normalized_density_plot, fig.height = 10, fig.width = 10, fig.align = "center"}
plotDensities(norm_data, log = TRUE, legend = "topright" , main = "Density of normalized intensities for multiple arrays")
```

- norm_data = EList object containing normalized intensities and information about the arrays
- log = Specifying the densities should be plotted on the log2 scale
- legend = Specifying where the legend should be placed on the graph
- main = title of the graph

```{r Normalized_MA_Plots}
count = 0 
for (sample_name in colnames(raw_data$E)) {
    count = count + 1
    limma::plotMA(norm_data, array = count, xlab = "Average Log-expression", ylab = "Expression Log-ratio (this sample vs. others)", main = sample_name)
}
```

- norm_data = EList object containing normalized intensities and information about the arrays
- array = choosing which array (sample) to use the data from
- xlab = label for the x axis
- ylab = label for the y axis
- main = title for the plot

```{r Normalized_data_boxplot}
boxplot(log2(raw_data$E)) # Comparing the raw data to the normalized data
suppressWarnings(boxplot(log2(norm_data$E)))
```

- raw_data$E = Raw intensities 
- norm_data$E = Normalized intensities

## Generating Probe and Gene Level Data
```{r Generating_probe_level_data}
# Generating Raw probe level data
raw_intensities_df <- as.data.frame(raw_data$E, row.names = raw_data$genes$GeneName, col.names = colnames(raw_data$E))# Generates a dataframe containing the raw intensities with the genes as the rows and samples as column names.
raw_probe_level_data_df <- cbind(raw_data$genes, raw_intensities_df)
write.csv(raw_probe_level_data_df, file = file.path(dir, "/output_data/GLDS-156_output_data/Raw_probe_level_data-GLDS-156.csv"), row.names = FALSE)
```

- raw_intensities_df = Variable containing the raw intensities (raw_data$E)
- row.names = Making the row names the names of the genes
- col.names = Making the column names the sample names
- raw_probe_level_data_df = Dataframe containing the raw intensities with information about each gene
- file = Choosing the output location of the csv file and the name for the file
- row.names = Telling R whether or not to export the row names into the csv file

```{r Generating_gene_level_data}
# Generating raw gene level data
first_sample_raw_mean = paste0(colnames(raw_data$E)[1], "_mean")
sample_one_raw = colnames(raw_data$E)[1]
summarized_columns_raw <- raw_probe_level_data_df %>% group_by(GeneName) %>% summarize(!!first_sample_raw_mean := mean(get(sample_one_raw))) # Using this data frame to merge with
for (sample_name_raw in colnames(raw_data$E)) {
    print(sample_name_raw)
    sample_name_mean = paste0(sample_name_raw,"_mean")
    sample_name_sd = paste0(sample_name_raw,"_sd")
    sample_name_median = paste0(sample_name_raw,"_median")
    summarized_columns_raw <- merge(summarized_columns_raw, raw_probe_level_data_df %>% group_by(GeneName) %>% summarize(!!sample_name_mean := mean(get(sample_name_raw)), !!sample_name_sd := sd(get(sample_name_raw)), !!sample_name_median := median(get(sample_name_raw)), .groups = "keep"))
    } # This is where I'm merging the other data frames to the first one, to generate 1 data frame with all of the columnes (Gene_name, sample_mean, sample_sd, sample_median)
write.csv(summarized_columns_raw, file = file.path(dir, paste0("/output_data/GLDS-156_output_data/Raw_gene_level_data-GLDS-156.csv")), row.names = FALSE)
```

- first_sample_raw_mean = Variable containing the name of the first sample followed by _mean. Used to name the column
- sample_one_raw = Variable containing the name of the first sample
- summarized_columns_raw = Dataframe that will contain all of the summarized gene level information (GeneName, mean, SD, median)
- sample_name_raw = Variable that contains the current sample
- sample_name_mean = Variable that holds the current sample name followed by _mean. Used to name the column.
- sample_name_sd = Variable that holds the current sample name followed by _sd. Used to name the column.
- sample_name_median = Variable that holds the current sample name followed by _median. Used to name the column.
- .groups = Tells the function what to do about the grouping structure from the dataframe that it is getting the information from

```{r Generating_normalized_probe_level_data}
# Generating Normalized probe level data
norm_intensities_df <- as.data.frame(norm_data$E, row.names = norm_data$genes$GeneName, col.names = colnames(norm_data$E))
norm_probe_level_data_df <- cbind(norm_data$genes, norm_intensities_df)
write.csv(norm_probe_level_data_df, file = file.path(dir, "/output_data/GLDS-156_output_data/Normalized_probe_level_data-GLDS-156.csv"), row.names = FALSE)
```

- norm_intensities_df = Variable containing the normalized intensities (norm_data$E)
- row.names = Making the row names the names of the genes
- col.names = Making the column names the sample names
- norm_probe_level_data_df = Dataframe containing the normalized intensities with information about each gene
- file = Choosing the output location of the csv file and the name for the file
- row.names = Telling R whether or not to export the row names into the csv file

```{r Generating_normalized_gene_level_data}
# Generating normalized gene level data
first_sample_norm_mean = paste0(colnames(norm_data$E)[1], "_mean")
sample_one_norm = colnames(norm_data$E)[1]
summarized_columns_norm <- norm_probe_level_data_df %>% group_by(GeneName) %>% summarize(!!first_sample_norm_mean := mean(get(sample_one_norm)))
for (sample_name_norm in colnames(norm_data$E)) {
    print(sample_name_norm)
    sample_name_mean = paste0(sample_name_norm, "_mean")
    sample_name_sd = paste0(sample_name_norm, "_sd")
    sample_name_median = paste0(sample_name_norm, "_median")
    summarized_columns_norm <- merge(summarized_columns_norm, norm_probe_level_data_df %>% group_by(GeneName) %>% summarize(!!sample_name_mean := mean(get(sample_name_norm)), !!sample_name_sd := sd(get(sample_name_norm)), !!sample_name_median := median(get(sample_name_norm)), .groups = "keep"))
    }
write.csv(summarized_columns_norm, file = file.path(dir, "/output_data/GLDS-156_output_data/Normalized_gene_level_data-GLDS-156.csv"), row.names = FALSE)
```

- first_sample_norm_mean = Variable containing the name of the first sample followed by _mean. Used to name the column
- sample_one_norm = Variable containing the name of the first sample
- summarized_columns_norm = Dataframe that will contain all of the summarized gene level information (GeneName, mean, SD, median)
- sample_name_norm = Variable that contains the current sample
- sample_name_mean = Variable that holds the current sample name followed by _mean. Used to name the column.
- sample_name_sd = Variable that holds the current sample name followed by _sd. Used to name the column.
- sample_name_median = Variable that holds the current sample name followed by _median. Used to name the column.
- .groups = Tells the function what to do about the grouping structure from the dataframe that it is getting the information from

```{r Filtering_Out_Controls_From Normalized_Data}

control_probes <- norm_data$genes$ControlType==1L
IsExpr <- rowSums(norm_data$other$gIsWellAboveBG > 0) > 4 # Removes probes that don't appear to be expressed
unmapped_probes <- raw_probe_level_data_df$GeneName==raw_probe_level_data_df$ProbeName
norm_data_filtered <- norm_data[!control_probes & !unmapped_probes & IsExpr, ]
```

- control_probes = Contains the rows of genes where the ControlType column == 1 
- IsExpr = Only contains the genes where the foreground is much greater than the background intensity. This means that the probe is expressed.
- unmapped_probes = Contains the probes where the GeneName is the same as the ProbeName column.
- norm_data_filtered = Filters out the rows with the control probes and unmapped probes while retaining the rows that are expressed.

## Differential Gene Analysis

```{r DGE_analysis}
number_of_probes = nrow(norm_data_filtered)
design <- model.matrix(~ 0 + levels)
fit <- lmFit(norm_data_filtered, design)

fit.groups <- colnames(fit$design)[which(fit$assign == 1)]
fit.index <-  which(levels(levels) %in% fit.groups)
fit.group.names <- gsub(" ", "_", sub(", ", "_", unique(GLDS_41_rs$`Factor Value[gravity]`)))

### Create Contrast Model
cat("\nCreating contrast model\n")
combos<-combn(fit.groups,2) # generate matrix of pairwise group combinations for comparison
combos.names<-combn(fit.group.names,2)
contrasts<-c(paste(combos[1,],combos[2,],sep = "-"),paste(combos[2,],combos[1,],sep = "-")) # format combinations for limma:makeContrasts
contrast.names <-c(paste(combos.names[1,],combos.names[2,],sep = "v"),paste(combos.names[2,],combos.names[1,],sep = "v")) # format combinations for output table file names

cont.matrix <- makeContrasts(contrasts = contrasts,levels=design)


contrast.fit <- contrasts.fit(fit, cont.matrix)
contrast.fit <- eBayes(contrast.fit)
# results<-decideTests(contrast.fit, method = "separate", adjust.method = "BH", p.value = 0.05, lfc = 0.5) # This line is not important
summary(decideTests(contrast.fit[,-1]))
result_table <- topTable(contrast.fit, number = number_of_probes) 
```

- number_of_probes = Variable that contains the number of rows in the norm_data_filtered variable.
- design = Variable containing the control and experimental groups (factor values) as levels (categorical variables). 
  - levels = Factor values stored as levels (categorical variables).
- fit = Variable containing the MArrayLM object created by lmFit. Contains the result of fitting gene-wise linear models to the normalized intensities / log ratios. 
  - norm_data_filtered = Normalized data where control probes, unmapped probes are removed and only the expressed probes are included.
- fit.groups = Variable containing the names of the levels of the experiment (control group and experimental groups). The names are in a format that is usable by R.
- fit.group.names = Variable containing the names of the levels of the experiment (control and experimental group). The names are not in a format that is usable by R (more human readable).
- combos = Matrix of the pairwise group combinations for comparison
- combos.names = Matrix of the pairwise group combinations for comparison using the group names that are more human readable.
- contrasts = Formatting the combinations to be used in the makeContrasts function.
- contrasts.names = Formatting the combinations to be used in the makeContrasts function using the group names that are more human readable and a 'v' rather than a dash to separate the pairwise combinations. 
- cont.matrix = Matrix containing pairwise combinations created by the makeContrasts function.
- contrast.fit = MArrayLM object containing the statistics (T-statistics, F-statistics, and log-odds of differential experession) as well as the contrasts (pairwise combinations). 
- results_table = Matrix that contains the genes that are significantly differentially expressed for each contrast. Contains the gene information, contrast information, and DGE information (P-values, Ave Expr, F, etc.).

## Adding Gene Annotations

```{r}
listEnsembl() 
ensembl <- useEnsembl(biomart = "genes") # Making a mart object (points to the data that is hosted on Ensembl website)
ensembl

searchDatasets(mart = ensembl, pattern = "Mouse") # Pattern = organism of interest
ensembl <- useDataset(dataset = "mmusculus_gene_ensembl", mart = ensembl) # Selecting the dataset

searchAttributes(mart = ensembl, pattern = "agilent_") # Look at comments of ADF file to figure out which attribute to use

my_agilent_ids <- norm_data_filtered$genes$ProbeName
results <- getBM(
    attributes = c(
        "agilent_wholegenome_4x44k_v1", # Name of the array from the ADF file (Write a markdown file about how I found the array name)
        "ensembl_gene_id",
        "uniprot_gn_symbol",
        "go_id"
        ), filters = "agilent_wholegenome_4x44k_v1", 
        values = c(my_agilent_ids), 
        mart = ensembl)

unique_probe_ids <- results %>% group_by(agilent_wholegenome_4x44k_v1) %>%
                      summarise(
                        SYMBOL = toString(unique(uniprot_gn_symbol)),
                        ENSEMBL = toString(unique(ensembl_gene_id)),
                        GO_Ids = toString(unique(go_id))
                        )
                      
# Join tables
bound_columns <- bind_cols(norm_data_filtered$E, norm_data_filtered$genes$ProbeName) %>% rename("ProbeName" = paste0("...", ncol(norm_data_filtered)+1))
compiled_table <- result_table %>% left_join(unique_probe_ids, by = c("ProbeName" = "agilent_wholegenome_4x44k_v1")) %>% left_join(bound_columns) 
write.csv(compiled_table, file = file.path(dir, "/output_data/GLDS-156_output_data/Probe_level_annotated_DGE_with_Normalized_Intensities_GLDS_156.csv"), row.names = FALSE) 
```

- listEnsembl = Lists the databases that Ensembl has.
- ensembl = Variable containing the mart object that points to the data that is hosted on the Ensembl website.
  - biomart = "genes" = Specifying that we want to use the gene database from Ensembl.
- searchDatasets = Lists out the datasets available in the database that follow the specific pattern indicated
  - pattern = "Mouse" = Specifying which organism dataset we want to look for in the database.
- dataset = "mmusculus_gene_ensembl" - Choosing the dataset from the list of available datasets for that organism.
- searchAttributes = A function that looks for the attribute that is specified in the pattern argument.
  - pattern = "agilent_" = Making the function look for attributes that contain "agilent_" in them.
- bound_columns = Matrix that links the normalized intensity values to their corresponding probe name.
  - rename("ProbeName" = paste0("...", ncol(norm_data_filtered)+1)) - Replacing the R generated column name with "ProbeName".
- compiled_table = Table that contains the merged DGE, normalized intensites, and gene annotations.
  - by = c("ProbeName" = "agilent_wholegenome_4x44k_v1") - Telling the function left_join how to join the two tables. In this case we're telling the function to use the ProbeName column from result_table and the "agilent_020186" column (probe names) from unique_probe_ids. 
- write.csv = Creates a CSV file containing the DGE, gene annotations, and normalized intensities into the specified output file.
